[CONJ-249] - Added attribute "enabledSslProtocolSuites" for manual setting of SSL protocol

krasaee opened the pull request:
I tested this with 10.1.14-MariaDB and it appears to work.. 

Please include in 1.4.6

Thanks.

- Kas

zhanhb commented:
mariadb 1.4.x requires at least jdk 1.7

rusher commented:
to follow this PR : this correspond to task https://jira.mariadb.org/projects/CONJ/issues/CONJ-249.
resumate : 
- no need to test java version, since 1.4.x Driver need 1.7 version that has all TLS version.
- best to activate only "TLSv1", "TLSv1.1" by default.  (catching exception for TLSv1.2 when server compile with yaSSL and relaunch with only "TLSv1", "TLSv1.1" would increase server failure stats without good reasons)
- TLSv1.2 has to be enabled only if parameter is set ( a new parameter )

krasaee commented:
@rusher I guess you did not pull this patch into the 1.4.6 release? do you want me to make any changes to it?

rusher commented on src/main/java/org/mariadb/jdbc/internal/util/JavaVersion.java:
```diff
@@ -0,0 +1,157 @@
+/**
+ * Created by krasaee on 2016-06-05.
+ */
+public class JavaVersion {
```
Unused class. Would be better to remove it

rusher commented on src/main/java/org/mariadb/jdbc/internal/protocol/AbstractConnectProtocol.java:
```diff
@@ -127,6 +128,11 @@
+    public static final String TLSv1 = "TLSv1";
+    public static final String TLSv11 = "TLSv1.1";
+    public static final String TLSv12 = "TLSv1.2";
+    public static final String TLSv13 = "TLSv1.3"; // not yet supported - in draft mode.
```
unused variable

rusher commented on src/main/java/org/mariadb/jdbc/internal/protocol/AbstractConnectProtocol.java:
```diff
@@ -797,6 +806,112 @@ public int getMinorServerVersion() {
+        }
+
+        // minimum of MariaDB 10.0.15 for TLSv1.2 support
+        if (!versionGreaterOrEqual(10, 0, 15)) {
```
musn't that be "if (!versionGreaterOrEqual(10, 0, 15)) {" (without the "!")

krasaee replied:
yes - lol not sure what happened there, I'll fix and submit again for the millionth time..

rusher commented:
Hi @krasaee, yes, this has not been part of the 1.4.6, but can be on 1.5.0 for the end of the month.

Your current implementation is good, but as indicated, adding TLSv1.2 by default according to version is too dangerous : packages are compiled using openSSL now, but some servers are build from source using yaSSL. Driver will throw "Unknown-0-0 version" Exception if TLSv1.2 is in the protocol list. So same application will work with some server, and won't with same version of server compiled differently.

Recent Heartbleed problem have make administrator using yaSSL more. No need to force TLSv1.2 today. 
Adding TLSv1.1 protocol, and permitting using TLSv1.2 with the "enabledSslProtocolSuites" options would permit to avoid that without creating any new risk. 

Additionally, similar to other open source projects, the MariaDB Foundation needs to have shared ownership of all code that is included in the MariaDB distribution. The easiest way to achieve this is by submitting your code under the BSD-new license. (The other alternative is to sign the code contribution agreement which can be found here: https://mariadb.com/kb/en/mariadb/mca/)

Please indicate in a comment below that you are contributing your new code of the whole pull request under the BSD-new license or that you have filled out the contribution agreement and sent it.

Thanks

krasaee commented:
@rusher I removed the unused JavaVersion util and changed default TLSv1.2 behaviour, forcing it to TLSv1.1 and TLSv1 (comments added to code). 

In respect to code contribution and licensing agreements, I agree to have all existing and future pull requests and code submissions be in accordance to the BSD-new license. In addition, I will submit the contribution agreement for any future code submitted outside this pull request.

krasaee commented:
@rusher closing pull request and reopening to start travis again (does not seem allow me to restart the build), keeps failing randomly..

krasaee commented:
@rusher all checks out with travis

rusher commented:
PR merged in 1.5 branch. 
closing this, since merge with branch doesn't close automatically PR
